mgmt: mcumgr: accept "offset reverted" packets to improve robustness

JiapengLi opened the pull request:
`reverted offset` could be caused by ble connection is unstable or mcu high load. Overwrite `fs_mgmt_ctxt.off` to `off` to continue upload process, this improves upload success rate

JiapengLi committed:
mgmt: mcumgr: accept "offset reverted" packets to improve robustness

`reverted offset` could be caused by ble connection is unstable
or mcu high load. Overwrite `fs_mgmt_ctxt.off` to `off` to
continue upload process, this improves upload success rate

Signed-off-by: JiapengLi <mail@jiapeng.me>

nordicjm reviewed (changes requested):
No, this is not valid, the MCUmgr client should send a command then wait for a response. Bluetooth is a reliable transport, so if a packet is dropped, it will be resent until it is either acknowledged or the supervision timeout occurs, at which point the connection will be dropped. For UDP it is a lossy transport so needs to deal with retries, for UART it should be reliable, and if not, enable hardware flow control. For other transports, same concept applies, client sends packet, waits for a response, if it gets no response after a period of time, try again (up to a maximum number of attempts) then if still no response, error out

JiapengLi commented:
@nordicjm Thank you for your prompt response. 

i am not sure how ble retry behaves, but i indeed test under ble link.
reverted offset also happens even if mcumgr client send the same packet at application layer. 
As I mentioned in the PR, reverted offset happens only at high mcu load (possible) or unstable BLE connection (may not)


my test log is like this:
 ```
                      off      fs_mgmt_ctxt.off
...
fs_mgmt_file_upload 285552 <-> 285552
fs_mgmt_file_upload 287972 <-> 287972
fs_mgmt_file_upload 290392 <-> 290392
# reverted offset
fs_mgmt_file_upload 285552 <-> 292812
fs_mgmt_file_upload 285552 <-> 292812
fs_mgmt_file_upload 285552 <-> 292812

# seesion aborted
 ```

my test environment:
- zephyr v3.2-branch
- nrf52840
- nRF Connect Device Manager v1.6.0

nordicjm commented:
You'd need to get a trace with e.g an nRF52840 loaded with the sniffer firmware showing where this "revert" is coming from, zephyr should not be sending a lower offset.

JiapengLi commented:
The reverted offset is sent by app, this is related to client upload api. 

OK. Let me try to sniff rf packets, and provide more details.

nordicjm commented:
@dinesharjani this sounds like an app bug?

dinesharjani commented:
@nordicjm this is Android? I think? The version number looks like Android. But I would not discount that this issue could be on the side of iOS as well, I can have a look at the code. Anyway, pinging @philips77

JiapengLi commented:
Yes. I tested through android nRF Connect Device Manager v1.6.0.

de-nordic reviewed (changes requested):
It should not be possible to get random packet sequence via bluetooth, unless client app breaks something.
Btw. This part of code is not bluetooth specific.

JiapengLi commented:
i will provide more information. Please allow me some time.

In my point of view, the transport layer can't always guarantee each pkt will be received only once, from my experience it is not easy, and costive. (For example mqtt QoS 2: exactly once.)
For current implementation, even a retry packet would stop the process (of course you can say it is on transport, but it is not always the way).

I do agree my case is not a general case. This patch just improve robustness, and doesn't break anything. It is some sort none common sense, but it do good.

nordicjm commented:
> In my point of view, the transport layer can't always guarantee each pkt will be received only once, from my experience it is not easy, and costive. (For example mqtt QoS 2: exactly once.)

Well in bluetooth it does actually, MQTT does too. In fact they both work similarly, for TCP/IP the host sends the data, the receiving device then sends an ACK packet - if the ACK packet is not received then the message will be sent again - it will be marked as a retransmission and have the same IP header details as the previous packet, so if the receiving device gets it again, it knows that the ACK packet was lost and will retransmit it, it will not give that data to the application again. Likewise in BLE, _all_ transmissions are acknowledged, the difference is in where they are acknowledged, indications and write with response packets are acknowledged at the application level, notifications and write without response packets are acknowledged at the stack level - the difference being that if the applications buffers or configuration is such that it cannot accept the data, notifications/write without response data will be dropped whereas indication/write with response will not and will need to be resent until it is accepted. If that is the issue you are seeing, then chances are your devices buffer configuration is inadequate for the amount of data the phone is sending, but that would not explain why the file write position counter would be going backwards.
